app/src/main/java/com/rpfcoding/snoozeloo/feature_alarm/scheduler_receiver/ReminderActivity.kt

```diff
@@ -0,0 +1,140 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.koin.android.ext.android.inject
+import org.koin.androidx.viewmodel.ext.android.viewModel
+import com.rpfcoding.snoozeloo.core.domain.ringtone.RingtoneManager as MyRingtoneManager
+
+class ReminderActivity : ComponentActivity() {
+
+    private val viewModel: ReminderViewModel by viewModel()
+    private val ringtoneManager: MyRingtoneManager by inject()
+    private val alarmScheduler: AlarmScheduler by inject()
+    private val scope: CoroutineScope by inject()
+    private val vibrator by lazy {
+        getSystemService(Vibrator::class.java)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        actionBar?.hide()
+        showOverLockscreen()
+        val alarmId = intent?.getStringExtra(AlarmConstants.EXTRA_ALARM_ID) ?: throw Exception("Alarm ID is not found.")
+
+        setContent {
+            SnoozelooTheme {
+                var effectsSet by remember { mutableStateOf(true) }
+
+                LaunchedEffect(Unit) {
+                    viewModel.getAlarmById(alarmId)
+                    delay(500L)
+                    effectsSet = false
+                }
+
+                LaunchedEffect(Unit) {
+                    delay(ALARM_MAX_REMINDER_MILLIS)
+                    viewModel.alarm?.let {
+                        disableAlarmAndFinish(it, it.isOneTime, shouldSnooze = true)
+                    }
+                }
+
+                LaunchedEffect(viewModel.alarm, effectsSet) {
+                    if (!effectsSet && viewModel.alarm != null) {
+                        effectsSet = true
+                        setupEffects(viewModel.alarm!!)
+                    }
+                }
+
+                viewModel.alarm?.let { alarm ->
+                    AlarmTriggerScreen(
+                        alarm = viewModel.alarm!!,
+                        onTurnOffClick = {
+                            disableAlarmAndFinish(alarm, alarm.isOneTime)
+                        },
+                        onSnoozeClick = {
+                            disableAlarmAndFinish(alarm, alarm.isOneTime, shouldSnooze = true)
+                        }
+                    )
+                }
+            }
+        }
+    }
+
+    private fun disableAlarmAndFinish(alarm: Alarm, isOneTime: Boolean, shouldSnooze: Boolean = false) {
+        if (shouldSnooze) {
+            alarmScheduler.schedule(
+                alarm = alarm,
+                shouldSnooze = true
+            )
+        } else {
+            if (isOneTime) {
+                viewModel.disableAlarm(alarm.id)
+            } else {
+                viewModel.rescheduleAlarm()
+            }
+        }
+        hideNotification(alarm.id.hashCode())
+        ringtoneManager.stop()
+        vibrator.cancel()
+        finish()
+    }
+
+    private fun setupEffects(alarm: Alarm) {
```

Comment:
All these functions live in Activity as functions that the UI calls.
The VM exists, but only has single function to reschedule alarms.

I’d comment on lack of consistency here. UI —> call VM functions